voter/src/main/java/org/terracotta/voter/ActiveVoter.java

```diff
@@ -117,169 +125,184 @@ public void awaitRegistrationWithAll(long timeout, TimeUnit unit) throws Interru
+              }
+            } catch (TimeoutException e) {
+              voterManager.close();
             }
-          } else {
-            LOGGER.info("State of {}: {}. Continuing the search for an active server.", voterManager.getTargetHostPort(), serverState);
           }
-        } catch (TimeoutException e) {
-          voterManager.close();
+          try {
+            Thread.sleep(REG_RETRY_INTERVAL);
+          } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+          }
         }
-      }
-    }, 0, REG_RETRY_INTERVAL, TimeUnit.MILLISECONDS)).collect(Collectors.toList());
-
+      }));
+    });
+    threads.forEach(thread -> thread.start());
     LOGGER.info("{} waiting to get registered with the active", this);
     try {
       return registrationLatch.get();
     } catch (ExecutionException e) {
       throw new RuntimeException(e);
     } finally {
-      futures.forEach(f->f.cancel(true));
+      threads.forEach(thread -> thread.interrupt());
     }
   }
 
-  public void registerAndHeartbeat(ExecutorService executorService, ClientVoterManager currentActive,
-                                   List<ClientVoterManager> voterManagers, Optional<Properties> connectionProps,
-                                   Set<String> registrationLatch) throws InterruptedException {
-    AtomicReference<ClientVoterManager> voteOwner = new AtomicReference<>();
-    //Try to connect and register with all the servers
-    voteOwner.set(currentActive);
-    List<Future<?>> futures = voterManagers.stream().map(voterManager -> executorService.submit(() -> {
-      try {
-        ClientVoterManager owner = voteOwner.get();
-        while (owner != null && !executorService.isShutdown()) {          
-          try {
-            voterManager.connect(connectionProps);
-            long lastVotedElection = 0;
-            // if the current vote owner is the active, allowed to reconnect
-            long registration = voterManager.registerVoter(id);
-            LOGGER.debug("Registration {} for {}", registration, voterManager);
-            if (registration > 0) {
-              if (voterManager == owner) {
-            // the vote was re-registered on the active, this is not allowed, voting cycle 
-            // must start over
-                if (voteOwner.compareAndSet(owner, null)) {
-                  voterManager.deregisterVoter(id);
-                }
-                voterManager.close();
-              } else {
-                if (owner.isVoting()) {
-                  voterManager.deregisterVoter(id);
+  private void heartbeat(ClientVoterManager voterManager,
+                         Optional<Properties> connectionProps,
+                         AtomicReference<ClientVoterManager> voteOwner) {
+    heartbeatThreads.put(voterManager.getTargetHostPort(), new Thread(() -> {
```

Comment:
may be we can use cached thread pool here and update the `heartbeatThreads` from the runnable to stop the thread when a server gets removed from topology